### PR TITLE
TW-1780: Open screen can connect with Twake Workplace when add another account

### PR DESCRIPTION
**Pull request by nqhhdev**
### Ticket

 - #1780 

### Need merge

https://github.com/linagora/linagora-design-flutter/pull/42

### Resolved

https://github.com/linagora/twake-on-matrix/assets/99852347/07dd61c1-0c99-4232-8e37-f1b4cfeb072c

### CODE REVIEW

File: lib/pages/twake_welcome/twake_welcome.dart [hunk 1]
```diff
@@ -58,8 +61,10 @@ class TwakeWelcomeController extends State<TwakeWelcome> with ConnectPageMixin {
 
   MatrixState get matrix => Matrix.of(context);
 
-  void onClickSignIn() {
+  void onClickSignIn() async {
     Logs().d("TwakeIdController::onClickSignIn: Login Url - $loginUrl");
+    final homeserverExisted = await _homeserverExisted();
+    if (homeserverExisted) return;
     _redirectRegistrationUrl(loginUrl);
```
Comment (hieutbui) [inline lib/pages/twake_welcome/twake_welcome.dart:68]: Found it same with logic inside `onClickCreateTwakeId()`. How about reuse?

File: lib/pages/twake_welcome/twake_welcome.dart [hunk 2]
```diff
@@ -79,15 +84,40 @@ class TwakeWelcomeController extends State<TwakeWelcome> with ConnectPageMixin {
...
   }
 
+  Future<bool> _homeserverExisted() async {
+    if (widget.arg?.isAddAnotherAccount == false) return false;
+    final allHomeserverLoggedIn = (await ClientManager.getClients())
+        .map((client) => client.homeserver.toString())
+        .toList();
+    Logs().i('All homeservers: $allHomeserverLoggedIn');
+    final homeserverExists = allHomeserverLoggedIn.any(
+      (homeserver) => "$homeserver/".contains(AppConfig.homeserver),
```
Comment (Te-Z) [inline lib/pages/twake_welcome/twake_welcome.dart:103]: can you explain this please ?

File: lib/pages/twake_welcome/twake_welcome.dart (hunk 1 above)
Comment (nqhhdev) [reply lib/pages/twake_welcome/twake_welcome.dart]: Updated it

File: lib/pages/twake_welcome/twake_welcome.dart (hunk 2 above)
Comment (nqhhdev) [reply lib/pages/twake_welcome/twake_welcome.dart:101]: When use add another account with TWP. I get all current `homeserver` and  check if `homeserver` existed and logged.
  - If `homeserver` existed → Can't support login 2 accounts with 1 `homeserver`
   example: `homeserver` logged = ['domain1', 'domain2'] and click on button login with 'domain1' → return error
  - If `homeserver` does not exist → continue

**Review by hieutbui — APPROVED** [review]

File: pubspec.lock [hunk 3]
```diff
@@ -1533,8 +1533,8 @@ packages:
     dependency: "direct main"
     description:
       path: "."
-      ref: master
-      resolved-ref: "7ed486ea00f7e91254555a4a828ff3a625b4e1ae"
+      ref: add-app-bar-for-twake-welcome
```
Comment (hoangdat) [inline pubspec.lock:1536]: please update it

File: pubspec.yaml [hunk 4]
```diff
@@ -29,7 +29,7 @@ dependencies:
   linagora_design_flutter:
     git:
       url: git@github.com:linagora/linagora-design-flutter.git
-      ref: master
+      ref: add-app-bar-for-twake-welcome
```
[the remark this answers is not in the document]
Comment (nqhhdev) [reply pubspec.yaml]: Done

**Review by Te-Z — APPROVED** [review]

File: lib/config/go_routes/go_router.dart [hunk 5]
```diff
@@ -375,7 +375,9 @@ abstract class AppRoutes {
               redirect: loggedOutRedirect,
               pageBuilder: (context, state) => defaultPageBuilder(
                 context,
-                const HomeserverPicker(),
+                TwakeWelcome(
+                  arg: state.extra as TwakeWelcomeArg?,
+                ),
```
Comment (sherlockvn) [inline lib/config/go_routes/go_router.dart:380]: checking state.extra is `TwakeWelcomeArg`

File: lib/pages/twake_welcome/twake_welcome.dart [hunk 6]
```diff
@@ -79,15 +84,38 @@ class TwakeWelcomeController extends State<TwakeWelcome> with ConnectPageMixin {
...
     );
     _redirectRegistrationUrl(signupUrl);
   }
 
+  Future<bool> _homeserverExisted() async {
+    if (widget.arg?.isAddAnotherAccount == false) return false;
+    final allHomeserverLoggedIn = (await ClientManager.getClients())
+        .map((client) => client.homeserver.toString())
+        .toList();
+    Logs().i('All homeservers: $allHomeserverLoggedIn');
```
Comment (sherlockvn) [inline lib/pages/twake_welcome/twake_welcome.dart:99]: better if we wrap it in try/catch

File: lib/pages/twake_welcome/twake_welcome.dart [hunk 7]
```diff
@@ -58,12 +61,14 @@ class TwakeWelcomeController extends State<TwakeWelcome> with ConnectPageMixin {
 
   MatrixState get matrix => Matrix.of(context);
 
-  void onClickSignIn() {
+  void onClickSignIn() async {
```
Comment (hoangdat) [inline lib/pages/twake_welcome/twake_welcome.dart:64]: remove

File: lib/pages/twake_welcome/twake_welcome.dart [hunk 8]
```diff
@@ -79,15 +84,43 @@ class TwakeWelcomeController extends State<TwakeWelcome> with ConnectPageMixin {
     handleTokenFromRegistrationSite(matrix: matrix, uri: uri);
   }
 
-  void onClickCreateTwakeId() {
+  void onClickCreateTwakeId() async {
```
Comment (hoangdat) [inline lib/pages/twake_welcome/twake_welcome.dart:87]: idem

File: lib/pages/twake_welcome/twake_welcome.dart [hunk 9]
```diff
@@ -79,15 +84,43 @@ class TwakeWelcomeController extends State<TwakeWelcome> with ConnectPageMixin {
...
       "TwakeIdController::onClickCreateTwakeId: Signup Url - $signupUrl",
     );
     _redirectRegistrationUrl(signupUrl);
   }
 
+  Future<bool> _homeserverExisted() async {
+    if (widget.arg?.isAddAnotherAccount == false) return false;
+    try {
+      final allHomeserverLoggedIn = (await ClientManager.getClients())
+          .map((client) => client.homeserver.toString())
```
Comment (hoangdat) [inline lib/pages/twake_welcome/twake_welcome.dart:98]: what happen if `homeserver == null`

File: lib/pages/twake_welcome/twake_welcome.dart [hunk 10]
```diff
@@ -86,8 +91,37 @@ class TwakeWelcomeController extends State<TwakeWelcome> with ConnectPageMixin {
     _redirectRegistrationUrl(signupUrl);
   }
 
+  Future<bool> _homeserverExisted() async {
+    if (widget.arg?.isAddAnotherAccount == false) return false;
```
Comment (sherlockvn) [inline lib/pages/twake_welcome/twake_welcome.dart:95]:
```suggestion
    if (widget.arg?.isAddAnotherAccount == false || widget.arg?.isAddAnotherAccount ==  null) return false;
```

**Review by hoangdat — APPROVED** [review]

**Pull request merged**